oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -63,98 +70,149 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
+                        worker.buildBatchRequests(innerPrepareRequests);
+                        synchronized (prepareRequests) {
+                            prepareRequests.addAll(innerPrepareRequests);
+                            if (prepareRequests.size() >= maxSyncoperationNum) {
+                                prepareRequests.notify();
+                            }
+                        }
+                        worker.endOfRound(System.currentTimeMillis() - lastTime);
+                    } finally {
+                        timer.finish();
+                        countDownLatch.countDown();
                     }
-
-                    worker.buildBatchRequests(prepareRequests);
-
-                    worker.endOfRound(System.currentTimeMillis() - lastTime);
                 });
+            });
+
+            Future<?> batchFuture = batchExecutorService.submit(() -> {
+                List<Future<?>> results = new ArrayList<>();
+                while (true) {
+                    synchronized (prepareRequests) {
+                        if (prepareDone && CollectionUtils.isEmpty(prepareRequests)) {
+                            break;
+                        }
+                    }
 
-                if (debug) {
-                    log.info("build batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
-                }
-            } finally {
-                timer.finish();
-            }
+                    synchronized (prepareRequests) {
+                        while (this.prepareRequests.size() < maxSyncoperationNum && !prepareDone) {
+                            try {
+                                this.prepareRequests.wait(1000);
+                            } catch (InterruptedException e) {
+                            }
+                        }
 
-            HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
-            try {
-                List<List<PrepareRequest>> partitions = Lists.partition(prepareRequests, maxSyncoperationNum);
-                CountDownLatch countDownLatch = new CountDownLatch(partitions.size());
-                for (final List<PrepareRequest> partition : partitions) {
-                    executorService.submit(() -> {
+                        if (CollectionUtils.isEmpty(prepareRequests)) {
+                            continue;
+                        }
+                    }
+
+                    List<PrepareRequest> partition = null;
+                    synchronized (prepareRequests) {
+                        List<PrepareRequest> prepareRequestList = this.prepareRequests.subList(0, Math.min(maxSyncoperationNum, this.prepareRequests.size()));
+                        partition = new ArrayList<>(prepareRequestList);
+                        prepareRequestList.clear();
+                    }
+                    List<PrepareRequest> finalPartition = partition;
```

Comment:
Why need `finalPartition`? It seems as same as `partition` always.

Reply:
> Why need `finalPartition`? It seems as same as `partition` always.

You need to make it `final` when passing it into a lambda

Reply:
Then let's keep a clear final.

Reply:
How about this ?
```
final List<PrepareRequest> finalPartition = partition;
```

Reply:
> How about this ?
> 
> ```
> final List<PrepareRequest> finalPartition = partition;
> ```

Yes, that's what we mean